Add promotion, singling, and deriving for Show

RyanGlScott opened the pull request:
Now that `GHC.TypeLits` features an `AppendSymbol` type family, it's possible to promote/single 90% of the `Show` type class and its features. (For the 10% that isn't covered, see my comment above the `Show Symbol` instance, as well as the comment explaining the lack of a `PShow Nat` instance.)

This also adds support for `deriving Show`, along with some test cases to stress-test it.

goldfirere commented:
Wow. This is quite a bit of work!

I hate to ask, but: Why? It has occurred to me to allow this kind of thing, but I always thought a singletonized version of `Show` was useless.

RyanGlScott commented:
My use case is type-level error messages, and works towards improving the story for #160. Currently, you have only primitive ways of constructing an `ErrorMessage`:

```haskell
data {-kind-} ErrorMessage = Text Symbol
                           | forall t. ShowType t
```

`ShowType` gives you one way of lifting arbitrary types to error messages, but it locks you into a particular way of rendering the types. There are a number of things about `ShowType`'s pretty-printing that I (and imagine others) would find unsavory, such as:

1. `ShowType` always renders promoted ticks, which can look messy:

   ```
   λ> let f :: TypeError (ShowType (Just True)); f = undefined
   
   <interactive>:11:10: error:
       • 'Just 'True
       • In the type signature:
           f :: TypeError (ShowType (Just True))
   ```
2. `ShowType` doesn't take fixity into account:

   ```
   λ> let f :: TypeError (ShowType (True `Infix` True)); f = undefined
   
   <interactive>:18:10: error:
       • Infix 'True 'True
       • In the type signature:
           f :: TypeError (ShowType (True `Infix` True))
   ```
3. More generally, `ShowType` is not _customizable_. The only way to make `ShowType` give the output you want to is carefully craft your types such that GHC's pretty-printer displays them the way you want, but that approach doesn't scale.

So, the other option is to use `Text`. But _prima facie_, there appears to be only one way to use `Text`: construct `Symbol`s by hand. But not with a promoted `PShow` class! `PShow` grants you the power to come up with flexible ways of turning your types into `Symbol`s in a composable way.

To really solve #160, we'll probably want to distinguish between `showType` (an `ErrorMessage` constructor) and `show'` (what this PR provides), as well as `(<>:)` (an `ErrorMessage` constructor) and `(<>)` (what this PR provides), for folks who want to use both, but I decided that was out of the scope of this PR.

goldfirere commented:
A compelling answer. I will review shortly.

goldfirere commented on src/Data/Singletons/Deriving/Show.hs:
```diff
@@ -0,0 +1,140 @@
+
+    -- A record constructor with no fields? This probably won't ever happen,
+    -- but just to be safe, we'll act as if it were a DNormalC.
+    go (DRecC []) = go (DNormalC False [])
```
Yes, this can really happen. GHC's `Show` instance does not print the extra braces, just as yours behaves here.

RyanGlScott replied:
Ah yes. I thought that GHC turned `data Foo = Foo {}` into a `NormalC`, but that's only if you _reify_ it. `[d| data Foo = Foo {} |]` does get turned into `RecC`, as I just experimentally confirmed. I'll edit the comment accordingly.

goldfirere commented on src/Data/Singletons/Prelude.hs:
```diff
@@ -54,6 +54,9 @@ module Data.Singletons.Prelude (
   module Data.Singletons.Prelude.Num,
 
+  -- * Singleton 'Show'
+  module Data.Singletons.Prelude.Show,
```
Ditto my comment above about the "promotion" Prelude.

RyanGlScott replied:
Ditto my response.

goldfirere commented on src/Data/Singletons/Prelude/Show.hs:
```diff
@@ -0,0 +1,229 @@
+-----------------------------------------------------------------------------
+-- |
+-- Module      :  Data.Singletons.Prelude.Show
+-- Copyright   :  (C) 2013 Richard Eisenberg
```
You may as well put your own name here...

RyanGlScott replied:
I'll do that.

goldfirere commented on src/Data/Singletons/Prelude/Show.hs:
```diff
@@ -0,0 +1,229 @@
+
+import           Unsafe.Coerce (unsafeCoerce)
+
+-- | The promoted analogue of '(<>)' for 'Symbol's. This uses the special
```
I don't want to be forward-incompatible with `Data.Monoid`'s `(<>)`. But I suppose this isn't incompatible with it.

RyanGlScott replied:
That's a good point. What we should _really_ be doing is defining a `Monoid` instance for `Symbols` first, and then use `Data.Monoid.<>` instead of defining it here. But we don't have a `Monoid` class in `singletons` yet, so I went with the simpler approach of rolling my own.

goldfirere commented on src/Data/Singletons/Prelude/Show.hs:
```diff
@@ -0,0 +1,229 @@
+  class Show a where
+    showsPrec :: Nat -> a -> SymbolS
+    -- Intentionally renamed to avoid clashing with the Show class
+    show'     :: a -> Symbol
```
In other places, `singletons` uses a trailing underscore to avoid naming conflicts. See `Data.Singletons.Prelude.Maybe.maybe_`. But remember to export the term-level name, then, too.

RyanGlScott replied:
I could rename `show'` to `show_`. But that presents a problem when actually defining. Defining `maybe_` is simple, since you can just use `singletons`:

```haskell

$(singletons [d|
  -- Renamed to avoid name clash
  maybe_ :: b -> (a -> b) -> Maybe a -> b
  maybe_ n _ Nothing  = n
  maybe_ _ f (Just x) = f x
 |])
```

But we can't do the same for `show_`, since it's a class method of `Show`, which we need to define with `singletonsOnly`, not `singletons`, as we don't want to define another `Show` class. Would you be alright with me just defining this locally?

```haskell
show_ :: Show a => a -> String
show_ = Prelude.show
```

goldfirere replied:
Yes, that's what I meant. Thanks.

goldfirere replied:
Good grief that's complex. I was just thinking of
```
-- works only for non-negatives
toString :: Integer -> String
toString 0 = "0"
toString 1 = "1"
toString 2 = "2"
...
toString 9 = "9"
toString n = toString (n `div` 10) <> toString (n `mod` 10)
```

But you don't have to if that runs into any trouble.

goldfirere replied:
Oops. Commented in the wrong spot, clearly.

goldfirere commented on src/Data/Singletons/Prelude/Show.hs:
```diff
@@ -0,0 +1,229 @@
+          = show_tuple [shows a, shows b, shows c, shows d, shows e, shows f, shows g] s
+  |])
+
+-- At the moment, I can't think of a way to implement a PShow Nat instance,
```
It can't be that hard... just write a normal `toString :: Integer -> String` and promote it, no? :)

RyanGlScott replied:
I was afraid you'd ask that :)

I suppose I could puzzle my way to an implementation, if you'd want that. I opted not to in this iteration since the [value-level `Integer -> String`](https://github.com/ghc/ghc/blob/ba597c1dd1daf9643b72dc7aeace8d6b3fce84eb/libraries/base/GHC/Show.hs#L471-L548) function that GHC implements is too complicated to promote, and I'm not clever enough to figure out how to make a simpler version off the top of my head.

goldfirere replied:
Good grief that's complex. I was just thinking of
```
-- works only for non-negatives
toString :: Integer -> String
toString 0 = "0"
toString 1 = "1"
toString 2 = "2"
...
toString 9 = "9"
toString n = toString (n `div` 10) <> toString (n `mod` 10)
```

But you don't have to if that runs into any trouble.

RyanGlScott replied:
How would you promote `div` and `mod`? Both of them also rely on primitive `integer-gmp` operations.

goldfirere replied:
See, for example, https://coq.inria.fr/library/Coq.Init.Nat.html

vagarenko replied:
Why `Div` and `Mod` are not implemented in GHC.TypeLits?
I asked some time ago on trac https://ghc.haskell.org/trac/ghc/ticket/13652 , but no one answered.

RyanGlScott replied:
@vagarenko, the answer to the question "why isn't this in `base`?" is almost always "because you haven't implemented it yet", and that holds true here as well.

That being said, Trac is not the place to make requests to add things to the standard libraries. That's better suited to a discussion on the [libraries mailing list](https://mail.haskell.org/mailman/listinfo/libraries).

goldfirere commented on src/Data/Singletons/Prelude/Show.hs:
```diff
@@ -0,0 +1,229 @@
+    case ex of
+      SomeSymbol (_ :: Proxy s) -> unsafeCoerce (SSym :: Sing s)
+
+$(singShowInstances [ ''(), ''Maybe, ''Either, ''NonEmpty, ''Bool, ''Ordering ])
```
Lists? Or is that handled by `showList`?

RyanGlScott replied:
I manually defined a `Show [a]` instance above. I didn't derive it since lists get shown specially.

RyanGlScott commented on src/Data/Promotion/Prelude.hs:
```diff
@@ -42,6 +42,9 @@ module Data.Promotion.Prelude (
   module Data.Promotion.Prelude.Num,
 
+  -- * Promoted 'Show'
+  module Data.Promotion.Prelude.Show,
```
Hm, good catch. I'll be more selective with what I export.

RyanGlScott replied:
A consequence of this, unfortunately, is that we have to add more `import Data.Singletons.Prelude.Show` statements to various test cases to work around #190. Oh well.

RyanGlScott commented on src/Data/Promotion/Prelude/Show.hs:
```diff
@@ -0,0 +1,49 @@
+-----------------------------------------------------------------------------
+
+module Data.Promotion.Prelude.Show (
+  PShow(..), SymbolS, SChar,
```
Good catch. I'll do that.

RyanGlScott commented on singletons.cabal:
```diff
@@ -44,9 +44,10 @@ library
                       template-haskell,
                       containers >= 0.5,
-                      th-desugar >= 1.7 && < 1.8,
+                      th-desugar >= 1.7.1 && < 1.8,
```
Good thinking. I'll bump it up.

RyanGlScott commented on src/Data/Singletons/Deriving/Show.hs:
```diff
@@ -0,0 +1,140 @@
+
+mkShowInstance :: DsMonad q => DType -> [DCon] -> q UInstDecl
+mkShowInstance ty cons = do
+  when (null cons) $
```
Well, standalone deriving _can_ handle it, but it's perhaps not the implementation you'd expect:

```
λ> data Foo
λ> deriving instance Show Foo

==================== Derived instances ====================
Derived instances:
  instance GHC.Show.Show Ghci4.Foo where
    GHC.Show.showsPrec = GHC.Err.error "Void showsPrec"
```

We don't have the ability to promote `error` yet, so that's not an option. And I'm not sure what other implementation you have in mind for empty datatypes, so I left it as an error.

RyanGlScott commented on src/Data/Singletons/Deriving/Show.hs:
```diff
@@ -0,0 +1,140 @@
+
+mk_showsPrec :: DsMonad q => [DCon] -> q [DClause]
+mk_showsPrec cons = do
+    p <- newUniqueName "p" -- The precedence argument (not always used)
```
In the code for `mk_showsPrec_clause`, I use `DWildPa`s in place of `DVarPa p` whenever `p` isn't used, so there's no need to prefix it with an underscore.

RyanGlScott commented:
I've added a dog-slow `PShow Nat` instance, per your request. As @vagarenko, it could be a lot more efficient if we had built-in `Divmod`, `Div`, and `Mod` type families in `GHC.TypeNats`, but that's not an option at the moment.

I believe I've addressed all of your comments. Let me know if you'd like any other changes.

goldfirere commented:
Thanks -- merged.
